src/System.Collections/src/System/Collections/BitArray.cs

```diff
@@ -218,18 +230,21 @@ public bool Get(int index)
         =========================================================================*/
         public void Set(int index, bool value)
         {
-            if (index < 0 || index >= Length)
+            if ((uint)index >= (uint)Length)
             {
                 throw new ArgumentOutOfRangeException(nameof(index), index, SR.ArgumentOutOfRange_Index);
             }
 
+            int quotient = index >> BitShiftPerInt32;   // Divide by 32.
+            int remainder = index - (quotient << BitShiftPerInt32);   // Multiply by 32.
+
             if (value)
             {
-                m_array[index / 32] |= (1 << (index % 32));
+                m_array[quotient] |= 1 << remainder;
```

Comment:
For the use case that triggered all these changes this `Set` implementation may not be so good. It's very large so it doesn't inline and it makes no difference that `value` was a constant - you still get a branch. What would probably be useful to add `Set(int index)` and `Reset(int index)` so the branch is not needed. But that's an API change.

Another thing that may not work so well is the use of `|=` and `&=`. Something like
```C#
        int mask = 1 << (index & 31);
        int bits = m_array[index >> 5];

        if (value)
        {
            bits |= mask;
        }
        else
        {
            bits &= ~mask;
        }

        m_array[index >> 5] = bits;
```
might work better than changing the array element directly.

And speaking of API changes, the sad reality is that `BitArray` will probably be always be broken performance wise, at least due to its `IEnumerator`-based enumerator.